pkg/eval/builtin_fn_io.go

```diff
@@ -194,6 +195,180 @@ func print(fm *Frame, opts printOpts, args ...interface{}) {
 	}
 }
 
+//elvdoc:fn printf
+//
+// ```elvish
+// printf $fmt $value...
+// ```
+//
+// Like [`print`](#print) but uses a formatting template to control how the
+// values appear in the output. An implicit newline is not added. You must
+// include a newline in the format (e.g., `printf "%s\n" yes`) or a string
+// value if a newline is desired in the output. This command writes its output
+// to the byte stream.
+//
+// See Go's [`fmt`](https://golang.org/pkg/fmt/#hdr-Printing) package for
+// details about the formatting verbs and the various flags that modify the
+// default behavior; e.g., left versus right justification. We explicitly
+// handle only strings and float64 data types. Printing other types is likely
+// to result in unexpected output (but not an exception). You can use the
+// numeric formatting verbs (e.g., `%d`, `%f`) with either a float64 or a
+// string that can be converted to a number. If you use a float64 value to
+// the `%s` verb it will first be converted to a string using the rules for
+// [numbers](language.html#number]).
```

Comment:
A more systematic way to think of the approach here is that all the verbs first tries to convert the corresponding argument to the relevant type. Namely:

* `%s` converts the argument to a string
* `%d` and other integer verbs convert the argument to an int, if possible
* `%f` and other float verbs convert the argument to a float, if possible